crates/bevy_ecs/src/query/error.rs

```diff
@@ -1,26 +1,93 @@
 use thiserror::Error;
 
-use crate::entity::Entity;
+use crate::{entity::Entity, world::unsafe_world_cell::UnsafeWorldCell};
 
 /// An error that occurs when retrieving a specific [`Entity`]'s query result from [`Query`](crate::system::Query) or [`QueryState`](crate::query::QueryState).
 // TODO: return the type_name as part of this error
-#[derive(Debug, PartialEq, Eq, Clone, Copy, Error)]
-pub enum QueryEntityError {
+#[derive(Clone, Copy)]
+pub enum QueryEntityError<'w> {
     /// The given [`Entity`]'s components do not match the query.
     ///
     /// Either it does not have a requested component, or it has a component which the query filters out.
-    #[error("The components of entity {0:?} do not match the query")]
-    QueryDoesNotMatch(Entity),
+    QueryDoesNotMatch(Entity, UnsafeWorldCell<'w>),
     /// The given [`Entity`] does not exist.
-    #[error("The entity {0:?} does not exist")]
     NoSuchEntity(Entity),
     /// The [`Entity`] was requested mutably more than once.
     ///
     /// See [`QueryState::get_many_mut`](crate::query::QueryState::get_many_mut) for an example.
-    #[error("The entity {0:?} was requested mutably more than once")]
     AliasedMutability(Entity),
 }
 
+impl<'w> std::error::Error for QueryEntityError<'w> {}
+
+impl<'w> std::fmt::Display for QueryEntityError<'w> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::QueryDoesNotMatch(entity, world) => {
+                write!(
+                    f,
+                    "The query does not match the entity {entity}, which has components "
+                )?;
+                format_archetype(f, world, entity)
+            }
+            Self::NoSuchEntity(entity) => write!(f, "The entity {entity} does not exist"),
+            Self::AliasedMutability(entity) => write!(
+                f,
+                "The entity {entity} was requested mutably more than once"
+            ),
+        }
+    }
+}
+
+impl<'w> std::fmt::Debug for QueryEntityError<'w> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::QueryDoesNotMatch(entity, world) => {
+                write!(f, "QueryDoesNotMatch({entity} with components ")?;
+                format_archetype(f, world, entity)?;
+                write!(f, ")")
+            }
+            Self::NoSuchEntity(entity) => write!(f, "NoSuchEntity({entity})"),
+            Self::AliasedMutability(entity) => write!(f, "AliasedMutability({entity})"),
+        }
+    }
+}
+
+fn format_archetype(
+    f: &mut std::fmt::Formatter<'_>,
+    world: UnsafeWorldCell<'_>,
+    entity: Entity,
+) -> std::fmt::Result {
+    // We know entity is still life
+    let entity = world
+        .get_entity(entity)
+        .expect("entity does not belong to world");
```

Comment:
If you expect it to exist, why not use `.entity()`? If it might not exist, why not return an error?